pkg/notify/notifier/pushover/pushover.go

```diff
@@ -0,0 +1,225 @@
+	timeout      time.Duration
+	logger       log.Logger
+	template     *notifier.Template
+	templateName string
+	client       http.Client
+}
+
+type pushoverRequest struct {
+	pushoverMessage
+}
+
+type pushoverResponse struct {
+	Status  int      `json:"status"`
+	Request string   `json:"request"`
+	Errors  []string `json:"errors,omitempty"`
+	Receipt string   `json:"receipt,omitempty"`
+}
+
+func NewPushoverNotifier(logger log.Logger, receivers []config.Receiver, notifierCfg *config.Config) notifier.Notifier {
+
+	var path []string
+	opts := notifierCfg.ReceiverOpts
+	if opts != nil && opts.Global != nil {
+		path = opts.Global.TemplateFiles
+	}
+	tmpl, err := notifier.NewTemplate(path)
+	if err != nil {
+		_ = level.Error(logger).Log("msg", "PushoverNotifier: get template error", "error", err.Error())
+		return nil
+	}
+
+	n := &Notifier{
+		notifierCfg:  notifierCfg,
+		timeout:      DefaultSendTimeout,
+		logger:       logger,
+		template:     tmpl,
+		templateName: DefaultTemplate,
+		client:       http.Client{Timeout: DefaultSendTimeout},
+	}
+
+	if opts != nil && opts.Pushover != nil {
+
+		if opts.Pushover.NotificationTimeout != nil {
+			n.timeout = time.Second * time.Duration(*opts.Pushover.NotificationTimeout)
+			n.client.Timeout = n.timeout
+		}
+
+		if len(opts.Pushover.Template) > 0 {
+			n.templateName = opts.Pushover.Template
+		} else if opts.Global != nil && len(opts.Global.Template) > 0 {
+			n.templateName = opts.Global.Template
+		}
+	}
+
+	for _, r := range receivers {
+		receiver, ok := r.(*config.Pushover)
+		if !ok || receiver == nil {
+			continue
+		}
+
+		if receiver.PushoverConfig == nil {
+			_ = level.Warn(logger).Log("msg", "PushoverNotifier: ignore receiver because of empty config")
+			continue
+		}
+
+		n.pushover = append(n.pushover, receiver)
+	}
+
+	return n
+}
+
+func (n *Notifier) Notify(ctx context.Context, data template.Data) []error {
+
+	send := func(userKey string, c *config.Pushover) error {
+
+		start := time.Now()
+		defer func() {
+			_ = level.Debug(n.logger).Log("msg", "PushoverNotifier: send message", "userKey", userKey, "used", time.Since(start).String())
+		}()
+
+		// retrieve app's token
+		token, err := n.notifierCfg.GetCredential(c.PushoverConfig.Token)
+		if err != nil {
+			_ = level.Error(n.logger).Log("msg", "PushoverNotifier: get token secret", "userKey", userKey, "error", err.Error())
+			return err
+		}
+
+		// filter data by selector
+		filteredData := utils.FilterAlerts(data, c.Selector, n.logger)
+		if len(filteredData.Alerts) == 0 {
+			return nil
+		}
+
+		// split new data along with its Alerts to ensure each message is small enough to fit the Pushover's message length limit
+		messages, _, err := n.template.Split(filteredData, MessageMaxLength, n.templateName, "", n.logger)
+		if err != nil {
+			_ = level.Error(n.logger).Log("msg", "PushoverNotifier: split alerts error", "userKey", userKey, "error", err.Error())
+			return err
+		}
+		for _, msg := range messages {
```

Comment:
This will send notifications sequentially, and the performance is relatively poor. You can refer to this code.
https://github.com/kubesphere/notification-manager/blob/master/pkg/notify/notifier/dingtalk/dingtalk.go#L386

Reply:
After a series of tests, sending messages in parallel will not be implemented. If the notifier attempts to send more than a certain amounts of messages at once, Pushover server will only allow these messages and respond with 429 Too Many Requests to the extra requests. Retrying failed requests is also impossible, since Pushover will send back 400 Bad Request to unchanged requests.

```
level=error ts=2021-07-31T02:48:10.6171142Z caller=pushover.go:191 msg="PushoverNotifier: got non-2xx response" userKey=** StatusCode=429 response="<html>\r\n<head><title>429 Too Many Requests</title></head>\r\n<body>\r\n<center><h1>429 Too Many Requests</h1></center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n"
// ... sleep
level=error ts=2021-07-31T02:48:11.8819076Z caller=pushover.go:191 msg="PushoverNotifier: got non-2xx response" userKey=** StatusCode=400 response="<html>\r\n<head><title>400 Bad Request</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<hr><center>cloudflare</center>\r\n</body>\r\n</html>\r\n"
// ... sleep
level=error ts=2021-07-31T02:48:12.9449241Z caller=pushover.go:191 msg="PushoverNotifier: got non-2xx response" userKey=** StatusCode=400 response="<html>\r\n<head><title>400 Bad Request</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<hr><center>cloudflare</center>\r\n</body>\r\n</html>\r\n"
```
Reference: https://pushover.net/api#friendly

Reply:
The pushover support to improve the message limit, so it’s necessary to send notifications in parallel.
https://support.pushover.net/i13-purchasing-additional-capacity-to-send-more-messages-per-month

There are two ways to deal with the message limits.

First, you can add a rate limiter to limit the frequency of notifications sending.
https://github.com/kubesphere/notification-manager/blob/master/pkg/notify/notifier/dingtalk/dingtalk.go#L400

Or, ignore this error when message limits have been reached if it will not lead to more serious consequences when you continue to send messages, such as prohibiting sending messages for a period of time.

Reply:
Resolved: https://github.com/txfs19260817/notification-manager/blob/pushover/pkg/notify/notifier/pushover/pushover.go#L145-L151